subsys/usb/usb_descriptor.c

```diff
@@ -818,8 +819,48 @@ void ascii7_to_utf16le(int idx_max, int asci_idx_max, u8_t *buf)
 	}
 }
 
+/*
+ * default USB SN string descriptor is CONFIG_USB_DEVICE_SN, but sometimes
+ * we want use another string as SN descriptor such as the chip's unique
+ * ID. So user can implement this function return a string thats will be
+ * replaced the default SN.
+ * please note that the new SN descriptor you changed must has same length
+ * as CONFIG_USB_DEVICE_SN.
+ */
+__weak u8_t *usb_set_runtime_ascii_sn_string_desc(void)
+{
+	return NULL;
+}
+
+static void usb_fix_ascii_sn_string_descriptor(void)
+{
+	u8_t *runtime_sn = usb_set_runtime_ascii_sn_string_desc();
+	int runtime_sn_len, default_sn_len;
+
+	if (!runtime_sn) {
+		return ;
+	}
+
+	/* both runtime and default SN's length doesn't include \0 */
+	runtime_sn_len = strlen(runtime_sn);
+	default_sn_len = sizeof(CONFIG_USB_DEVICE_SN) - 1;
+
+	if (runtime_sn_len != default_sn_len) {
+		SYS_LOG_ERR("the new SN descriptor doesn't has the same "
+			    "length as CONFIG_USB_DEVICE_SN");
+		for (; ;) {
+			/* Spin */
```

Comment:
Why? It should return errors or output warning, it is also not critical to the function of the stack.

Reply:
yeah it's better leaving a warring message and return, not enter a dead while.